### PR TITLE
lib/ClimaCoreSpectra

### LINKED ISSUE #967: Spectra for GCM configuration
## Purpose
Spectra on the sphere via spherical harmonics are a feature needed for diagnostic purposes. 

Link to any relevant PRs/issues: https://github.com/CliMA/ClimaDiagnostics/issues/1

Some of this diagnostic functionality existed in [`ClimateMachine`](https://github.com/CliMA/ClimateMachine.jl/tree/master/src/Common/Spectra) but will need to be ported over to ClimaCore and adapted to use its API/objects, and potentially optimized using existing Julia packages. 

## Cost/benefits/risks
Having spherical harmonics will allow us to calculate and plot the energy spectra to analyze turbulence or numerical dissipation (e.g., from hyperdiffusion). 

## Producers
Lead: @valeriabarra
Potential collaborators: @LenkaNovak , @akshaysridhar , @jiahe23 

## Components
On the sphere we want both 1D and 2D spectra. For these we need:
- Porting existing code:
  - [x] For 1D: Calculate the 1D (zonal) power spectra using the Fourier transform at each latitude and level from the 2D horizontal velocity field
  - [x] For 2D: Transform variable on lat-lon grid to the 2D spectral (spherical harmonic) space using `fft` on latitude circles (as for the 1D spectrum) and associated Legendre polynomials for meridians, and calculate spectra. 
- Enhancement
  - [ ] Reuse existing Julia libraries to favor library reuse to the extent possible, e.g., [`AssociatedLegendrePolynomials.jl`](https://github.com/jmert/AssociatedLegendrePolynomials.jl) for the calulation of the associated normalized Legendre polynomials and [`QuadGK.jl`](https://juliamath.github.io/QuadGK.jl/stable/) for Gaussian integration

## Inputs
References: 
- ClimateMachine [`Spectra`](https://github.com/CliMA/ClimateMachine.jl/tree/master/src/Common/Spectra) module and references therein:
    - Baer, ["An Alternate Scale Representation of Atmospheric Energy Spectra"](https://journals.ametsoc.org/view/journals/atsc/29/4/1520-0469_1972_029_0649_aasroa_2_0_co_2.xml), 1972
    - Ehrendorfer, ["Spectral Numerical Weather Prediction Models"](https://epubs.siam.org/doi/book/10.1137/1.9781611971996), 2011
     
## Results and deliverables

- [x] A ClimaCore subpackage that will live under `lib`, e.g., `lib/ClimaCoreSpectra`, which will include both library code under `src` and related tests under `test`, exercised through the usual `runtests.jl`. 
- [x] Unit test(s) 
  - [x] visual component for a specific spherical harmonic as the one in ClimateMachine did
  - [x] test against existing Python packages (e.g., https://shtools.github.io/SHTOOLS/complex-spherical-harmonics.html) (this was done in the original implementation)
  - [x] integration test / example that show the workflow: from a `Field` defined on the ClimaCore cubed-sphere to the regridding and specrta calculation  
- [x] Documentation in ClimaCore  

Integration in downstream repositories/packages:
- [ ] Implementation in ClimaCoupler diagnostics and visualization, featuring in the AMIP driver (I would open a separate issue for this in CCoupler for this)

## Task breakdown
A preliminary list of PRs and a preliminary timeline of PRs, milestones, and key results.
- [x] Development of the `lib/ClimaCoreSpectra` package [1 week?] 
- [x] unit testing
- [x] documentation
- [x] integration test/example that shows the full workflow (starting from a CC cubed-sphere, remapping to rll grid, and transforming to spectral space)

The four tasks above are implemented in https://github.com/CliMA/ClimaCore.jl/pull/979

Additional tasks for follow-up PRs: 
- [ ] Improved interface that uses ClimaCore `Fields` directly with internal calls to remapping infrastructure so to remove the burned from the user

Integration in downstream repositories/packages:
- [ ] add in AMIP driver (I would open a separate issue for this in CA for this. Test case to be used for initial integration test is [moist baroclinic wave](https://github.com/CliMA/ClimaAtmos.jl/blob/c28ea7a16a581c72a4fd7783f2744c011a04b46b/examples/hybrid/define_post_processing.jl#L323))
...

Details are subject to change as well as tasks/subtasks as things and plans are refined.

## Working branch
[`valeria/spherical-harmonics`](https://github.com/CliMA/ClimaCore.jl/tree/valeria/spherical-harmonics)

## Reviewers
@charleskawczynski and possibly other folks who have worked on/with the prior implementation (e.g., @LenkaNovak , @akshaysridhar )

### PR DESCRIPTION
**Pull request by valeriabarra**
This PR contains incremental work towards #967 

## Content: 

- [x] added a subpackage `lib/ClimaCoreSpectra` (with main implentation ported over from `ClimateMachine.jl` for now _not_ using ClimaCore's API object)
- [x] the implementation is type-safe and array type agnostic so that we could run this feature both on CPUs and GPUs (using `Array` or `CuArray`)
- [x] added and improved existing unit tests in `test/runtests.jl` and `test/gcm_visual_test.jl`
- [x] added and improved existing documentation, and linked a visual test to the `docs`
- [x] added a new integration test in `examples/gcm_remap_and_spectra.jl` that shows the entire workflow: starting from a ClimaCore cubed-sphere, defining a `Field`, remapping it using `ClimaCoreTempestRemap` and then generate the spectra with `ClimaCoreSpectra` calculator 
- [x] The plots generated and linked in the docs are tested to follow accessibility best practices for plotting (tested for color-blindness friendliness) 

- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [x] Documentation has been added/updated OR N/A.

### CODE REVIEW

File: lib/ClimaCoreSpectra/src/gcm_spectra.jl
Comment (valeriabarra) [inline lib/ClimaCoreSpectra/src/gcm_spectra.jl:15]: @charleskawczynski some of these remain open questions (for example, at the moment we only employ user-defined parameters and not CC API types/functions). I'd like to connect with you and @LenkaNovak to discuss some of these questions. Thanks

File: lib/ClimaCoreSpectra/src/gcm_spectra.jl
Comment (charleskawczynski) [inline lib/ClimaCoreSpectra/src/gcm_spectra.jl:223]:
```suggestion
    mesh::SpectralSphericalMesh{FT},
    pfield::AbstractArray,
) where {FT}
```

File: lib/ClimaCoreSpectra/src/gcm_spectra.jl
Comment (charleskawczynski) [inline lib/ClimaCoreSpectra/src/gcm_spectra.jl:255]:
```suggestion
                    ) .* qwg[m, n, 1:nθ_half] ./ 2
```

File: lib/ClimaCoreSpectra/src/gcm_spectra.jl
Comment (charleskawczynski) [inline lib/ClimaCoreSpectra/src/gcm_spectra.jl:227]:
```suggestion
        mesh.var_fourier[:, :, 1] * 0, mesh.var_spherical[:, :, 1, :] * 0
```

File: lib/ClimaCoreSpectra/src/gcm_spectra.jl
Comment (charleskawczynski) [inline lib/ClimaCoreSpectra/src/gcm_spectra.jl:320]:
```suggestion
            x = lon ./ 180 .* π
            dx = (lon[2] - lon[1]) ./ 180 .* π
```

File: lib/ClimaCoreSpectra/src/gcm_spectra.jl
Comment (charleskawczynski) [inline lib/ClimaCoreSpectra/src/gcm_spectra.jl:323]:
```suggestion
            freqs_ = fftfreq(num_fourier, 1 / dx) # 0,+ve freq,-ve freqs (lowest to highest)
            freqs[:, j, k] = freqs_[1:num_pfourier] .* 2 .* π
```

File: lib/ClimaCoreSpectra/src/gcm_spectra.jl
Comment (charleskawczynski) [inline lib/ClimaCoreSpectra/src/gcm_spectra.jl:337]:
```suggestion
                    2 * weight[k] * fourier[m] * conj(fourier[m]) # factor 2 for neg freq contribution
```

File: lib/ClimaCoreSpectra/src/gcm_spectra.jl
Comment (charleskawczynski) [inline lib/ClimaCoreSpectra/src/gcm_spectra.jl:389]:
```suggestion
        var_spectrum[1, :, k] = var_spectrum[1, :, k] ./ 2 # m=0
```

File: lib/ClimaCoreSpectra/test/gcm_visual_test.jl
Comment (charleskawczynski) [inline lib/ClimaCoreSpectra/test/gcm_visual_test.jl:31]: What is `var_grid`?

File: lib/ClimaCoreSpectra/src/gcm_spectra.jl
Comment (charleskawczynski) [reply lib/ClimaCoreSpectra/src/gcm_spectra.jl:15]: I think that these are the questions that need to be answered as a part of this work

File: .github/workflows/ClimaCoreSpectra.yml
Comment (valeriabarra) [reply .github/workflows/ClimaCoreSpectra.yml:1]: Ah, I followed the same convention that I've seen for the other subpackages, e.g., ClimaCoreTempestRemap, ClimaCoreVTK, etc.. Should this be different?

Comment (charleskawczynski) [reply .github/workflows/ClimaCoreSpectra.yml:1]: Ah, never-mind, I forgot that this is a shared folder, this name is fine

File: lib/ClimaCoreSpectra/test/gcm_visual_test.jl
Comment (valeriabarra) [reply lib/ClimaCoreSpectra/test/gcm_visual_test.jl:31]: It is the uniform (regular) lat-long grid on which we apply the transformation to the spectral space. Perhaps I should just rename this `rll_grid` as we have used in the remapping functionality

**Comment by LenkaNovak** [comment]
Hi Valeria, this looks great! As you say, plenty of clean up of the old code and docs is needed, but the structure/main functions are there. To close https://github.com/CliMA/ClimaCore.jl/issues/967 we should implement this in ClimaAtmos / ClimaCoupler. We may also need to think about interpolation from the RLL grid (output of TempestRemap) to Gaussian grid. Happy to chat offline later on.

File: lib/ClimaCoreSpectra/src/gcm_spectra.jl
Comment (charleskawczynski) [reply lib/ClimaCoreSpectra/src/gcm_spectra.jl:227]: It looks like you're timing in global scope. Please see https://docs.julialang.org/en/v1/manual/performance-tips/#Performance-critical-code-should-be-inside-a-function

Comment (valeriabarra) [reply lib/ClimaCoreSpectra/src/gcm_spectra.jl:227]: Thanks for the reminder. It seems like the change would not provide benefits then:
```
julia> function with_float() 
       var_fourier2d, var_spherical2d =
                      mesh.var_fourier[:, :, 1] * 0.0, mesh.var_spherical[:, :, 1, :] * 0.0
       end
with_float (generic function with 1 method)

julia> function with_int() 
       var_fourier2d, var_spherical2d =
                      mesh.var_fourier[:, :, 1] * 0, mesh.var_spherical[:, :, 1, :] * 0
       end
with_int (generic function with 1 method)

julia> @time with_float()
  0.000081 seconds (42 allocations: 318.094 KiB)
  
julia> @time with_int()
  0.000080 seconds (42 allocations: 318.094 KiB)
```
I'll be happy to discuss more offline

Comment (charleskawczynski) [reply lib/ClimaCoreSpectra/src/gcm_spectra.jl:227]: Let's make the changes I suggested and we can discuss the performance issues offline.

File: lib/ClimaCoreSpectra/src/gcm_spectra.jl
Comment (charleskawczynski) [inline lib/ClimaCoreSpectra/src/gcm_spectra.jl:31]:
```suggestion
struct SpectralSphericalMesh{FT, ArrF3, ArrI2, ArrC3, ArrC4} <:
               AbstractSpectralSphericalMesh{FT, ArrF3, ArrI2, ArrC3, ArrC4}
```

Comment (valeriabarra) [reply lib/ClimaCoreSpectra/src/gcm_spectra.jl:33]: On a second thought, the way this is set up now, it actually needs to be mutable. We mute one of its arguments later on. 
```
mesh.qnm = compute_legendre!(
        FT,
        mesh.num_fourier,
        mesh.num_spherical,
        sinθ,
        mesh.nθ,
    )
```
I would leave this to following clean-up PRs, as that `compute_legendre` would most likely be substitute by the AssociatedLegendrePolynomials.jl package functions

**Comment by valeriabarra** [comment]
@charleskawczynski , @LenkaNovak and I tested this extensively today and we verified that it gives the expected result. This is ready for final review. So that after this is merged, we can add integration of this spectra calculator in ClimaAtmos (which already uses the TempestRemap regridding functionality) and then later we can come back generalizing the interface so that it uses the CC `Fields`. 

Thanks for your review and help. 

I'd like to keep the commits to a minimum number of isolated commits that still preserve acknowledgement of co-authorship since people contributed :)  (so I wouldn't squash all commits to 1)

**Comment by valeriabarra** [comment]
My latest force-push to this branch was an attempt at rewording the commit messages to have GitHub pick up @LenkaNovak as a co-author of the last couple of commits. I don't know why it's not working. I am using the right syntax. I have tried both using her personal email address and the automatically generated "noreply" address. 
It has worked fine many other times for me but recently I was not picked up on someone else commit (in ClimaTimeSteppers.jl for instance) and now it is not picking up others on my commits here. I wonder if something has changed recently but I can't find any answer online

**Comment by LenkaNovak** [comment]
Don't worry about it, @valeriabarra , and thanks for pushing to be inclusive! :)

**Review by charleskawczynski — APPROVED** [review]

File: lib/ClimaCoreSpectra/Project.toml
Comment (charleskawczynski) [inline lib/ClimaCoreSpectra/Project.toml:12]: Let's add compat entries for the other deps

File: lib/ClimaCoreSpectra/test/gcm_visual_test.jl
Comment (charleskawczynski) [inline lib/ClimaCoreSpectra/test/gcm_visual_test.jl:8]:
```suggestion
OUTPUT_DIR = if haskey(ENV, "BUILD_DOCS")
    ""
else
    joinpath(@__DIR__, get(ENV, "CI_OUTPUT_DIR", tempname()))
end
mkpath(OUTPUT_DIR)
```

Comment (charleskawczynski) [reply lib/ClimaCoreSpectra/test/gcm_visual_test.jl:8]: Maybe try this?

**Pull request merged**
